tests/gradients/test_parameter_shift.py

```diff
@@ -1141,114 +1310,772 @@ def reference_qnode(x):
+        manualgrad_val = np.zeros_like(autograd_val)
 
-        tapes, fn = qml.gradients.param_shift(tape)
-        jac = fn(dev.batch_execute(tapes))
-        cost = jac[0, 0]
-        cost.backward()
-        hess = params.grad
+        for idx in list(np.ndindex(*params.shape)):
+            s = np.zeros_like(params)
+            s[idx] += np.pi / 2
 
-        x, y = params.detach().numpy()
+            tape.set_parameters(params + s)
+            forward = dev.execute(tape)
 
-        expected = np.array([np.sin(2 * x) * np.sin(y) ** 2, -np.cos(x) ** 2 * np.sin(2 * y)])
-        assert np.allclose(jac.detach().numpy(), expected, atol=tol, rtol=0)
+            tape.set_parameters(params - s)
+            backward = dev.execute(tape)
 
-        expected = np.array([2 * np.cos(2 * x) * np.sin(y) ** 2, np.sin(2 * x) * np.sin(2 * y)])
-        assert np.allclose(hess.detach().numpy(), expected, atol=0.1, rtol=0)
+            manualgrad_val[0, idx] = (forward - backward) / 2
 
-    @pytest.mark.jax
-    def test_jax(self, tol):
-        """Tests that the output of the finite-difference transform
-        can be differentiated using JAX, yielding second derivatives."""
-        import jax
-        from jax import numpy as jnp
-        from jax.config import config
+        assert np.allclose(autograd_val, manualgrad_val, atol=tol, rtol=0)
+        assert spy.call_args[1]["shifts"] == (shift,)
 
-        config.update("jax_enable_x64", True)
+        # compare to finite differences
+        tapes, fn = qml.gradients.finite_diff(tape)
+        numeric_val = fn(dev.batch_execute(tapes))
+        assert np.allclose(autograd_val, numeric_val, atol=tol, rtol=0)
 
-        dev = qml.device("default.qubit.jax", wires=2)
-        params = jnp.array([0.543, -0.654])
+    @pytest.mark.parametrize("G", [qml.CRX, qml.CRY, qml.CRZ])
+    def test_controlled_rotation_gradient(self, G, tol):
+        """Test gradient of controlled rotation gates"""
+        dev = qml.device("default.qubit", wires=2)
+        b = 0.123
+
+        with qml.tape.QuantumTape() as tape:
+            qml.QubitStateVector(np.array([1.0, -1.0], requires_grad=False) / np.sqrt(2), wires=0)
+            G(b, wires=[0, 1])
+            qml.expval(qml.PauliX(0))
+
+        tape.trainable_params = {1}
+
+        res = dev.execute(tape)
+        assert np.allclose(res, -np.cos(b / 2), atol=tol, rtol=0)
+
+        tapes, fn = qml.gradients.param_shift(tape, broadcast=True)
+        grad = fn(dev.batch_execute(tapes))
+        expected = np.sin(b / 2) / 2
+        assert np.allclose(grad, expected, atol=tol, rtol=0)
+
+        # compare to finite differences
+        tapes, fn = qml.gradients.finite_diff(tape)
+        numeric_val = fn(dev.batch_execute(tapes))
+        assert np.allclose(grad, numeric_val, atol=tol, rtol=0)
+
+    @pytest.mark.parametrize("theta", np.linspace(-2 * np.pi, np.pi, 7))
+    def test_CRot_gradient(self, theta, tol):

```

Comment:
What is the runtime of this test (can be checked via passing `--durations 0` in when invoking `pytest`)? We could get away with fewer param tests if it's significant.

Reply:
That's okay then :+1: Just arose in my mind because of choosing 7 args from the output of a `linspace`, but they are cheap enough. :100:

Reply:
Fair enough! :)